### PR TITLE
add private ip address for openstack driver

### PR DESCRIPTION
**Pull request by xinxian0458**
Signed-off-by: Zhang, Jie <zhangjie0220@gmail.com>

### CODE REVIEW

**Comment by nathanleclaire** [comment]
cc @ggiamarchi WDYT

**Comment by dgageot** [comment]
Hi @xinxian0458, I don't understand the purpose of this PR. Can you explain?

**Comment by xinxian0458** [comment]
@dgageot, In my environment, I need to know both the floating ip address and private ip address of the machine created in the openstack. but the current openstack driver only provide one ip address, I can't get its private ip address if i set the openstack-floatingip-pool.

File: drivers/openstack/openstack.go [hunk 1]
```diff
@@ -44,6 +44,7 @@ type Driver struct {
 	FloatingIpPoolId string
 	IpVersion        int
 	client           Client
+	PrivateIPAddress string
```
Comment (fsoppelsa) [inline drivers/openstack/openstack.go]: We already have `d.IPAddress`

File: drivers/openstack/openstack.go [hunk 2]
```diff
@@ -318,6 +319,35 @@ func (d *Driver) GetIP() (string, error) {
 	return "", fmt.Errorf("No IP found for the machine")
 }
 
+func (d *Driver) GetPrivateIP() (string, error) {
```
Comment (fsoppelsa) [inline drivers/openstack/openstack.go]: Did you test this in conjunction with a floating IP of `assignFloatingIP()`?

How does the command look like?

File: drivers/openstack/openstack.go [hunk 3]
```diff
@@ -318,6 +319,35 @@ func (d *Driver) GetIP() (string, error) {
...
+	log.Debug("Looking for the Private IP address...", map[string]string{"MachineId": d.MachineId})
+
+	if err := d.initCompute(); err != nil {
+		return "", err
+	}
+
+	addressType := Fixed
+
+	// Looking for the IP address in a retry loop to deal with OpenStack latency
+	for retryCount := 0; retryCount < 200; retryCount++ {
```
Comment (fsoppelsa) [inline drivers/openstack/openstack.go]: Where is `200` coming from?

File: drivers/openstack/openstack.go (hunk 1 above)
Comment (xinxian0458) [inline drivers/openstack/openstack.go]: Yes, I see. but in my opinion, the machine need its private ip no matter whether set assignFloatingIP or not, so it should has another attribute.

File: drivers/openstack/openstack.go (hunk 2 above)
Comment (xinxian0458) [inline drivers/openstack/openstack.go]: I just test it in the openstack which is using nova network.

docker-machine --debug create \
--driver openstack \
--openstack-auth-url http://172.19.0.2:5000/v2.0 \
--openstack-username demo \
--openstack-password password \
--openstack-tenant-name demo \
--openstack-flavor-id 4 \
--openstack-image-id 312a1b0c-e0b2-4968-a828-3e0b67c842d9 \
--openstack-ssh-user centos \
--openstack-sec-groups default \
--openstack-floatingip-pool public \
--openstack-nova-network=true \
openstack02

File: drivers/openstack/openstack.go (hunk 3 above)
Comment (xinxian0458) [inline drivers/openstack/openstack.go]: Just copy it from GetIP() function.

File: drivers/openstack/openstack.go (hunk 2 above)
Comment (fsoppelsa) [inline drivers/openstack/openstack.go]: Can you test that over Neutron too? (I suppose it should work)

**Comment by dgageot** [comment]
I'm really confused by this PR. I understand that it's goal is to add the machine's private ip in the `config.json`. The `config.json` is mainly an internal format and shouldn't be used by external tools.
So, until I better understand the goal, I`m going to NotLGTM

**Comment by fsoppelsa** [comment]
Easier, I would suggest:

Always use the private network.
If also a floating IP is needed, as the very latest step in machine's provisioning, allocate that additional port from the floating pool.

**Comment by dgageot** [comment]
@fsoppelsa If I understand correctly, there's a way to achieve @xinxian0458's goal without this PR. Right?
In that case, I'd like to close the PR. @fsoppelsa @xinxian0458 wdyt?

**Comment by fsoppelsa** [comment]
@dgageot well, my suggestion is to use the private network always for dm, and add some code in the very end to enable a floating IP.

This PR is not about this, it's adding things that IMHO are too much.

Getting the machine's floating or internal, is a matter of using the Nova APIs or even the CLI.

**Pull request closed without merging**